auraed/src/runtime/cell_service/executables/executable.rs

```diff
@@ -23,66 +38,179 @@ enum ExecutableState {
-            }
-            ExecutableState::Stopped(exit_status) => Ok(Some(*exit_status)),
+    pub fn kill(&mut self) -> io::Result<Option<ExitStatus>> {
+        let exit_status: Option<ExitStatus>;
+        {
+            let mut inner = self.inner.lock().map_err(|e| {
+                io::Error::new(io::ErrorKind::Other, e.to_string())
+            })?;
+            match &mut inner.state {
+                ExecutableState::Init { .. } => exit_status = None,
+                ExecutableState::Started { process, .. } => {
+                    let proc_status: ExitStatus;
+                    {
+                        let mut proc = process.lock().map_err(|e| {
+                            io::Error::new(io::ErrorKind::Other, e.to_string())
+                        })?;
+                        proc.kill(Some(SIGKILL))?;
+                        proc_status = proc.wait()?;
+                    }
+                    inner.state = ExecutableState::Stopped(proc_status);
+                    exit_status = Some(proc_status);
+                }
+                ExecutableState::Stopped(status) => exit_status = Some(*status),
+            };
         }
+        self.log_thread
+            .take()
+            .map(thread::JoinHandle::join)
+            .expect("thread panicked")
+            .expect("join log_thread")
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
+        Ok(exit_status)
     }
 
     /// Returns the [Pid] while [Executable] is running, otherwise returns [None].
-    pub fn pid(&self) -> Option<Pid> {
-        let ExecutableState::Started { process, .. } = &self.state else {
-            return None;
+    pub fn pid(&self) -> io::Result<Option<Pid>> {
+        let inner = self
+            .inner
+            .lock()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
+        let ExecutableState::Started { process, .. } = &inner.state else {
+            return Ok(None);
         };
+        let proc = process
+            .lock()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
 
-        Some(process.pid())
+        Ok(Some(proc.pid()))
     }
 }
 
+/// Spawns a thread that produces log lines while the [Executable] is running.
+fn spawn_log_thread(
+    inner: Arc<Mutex<ExecutableInner>>,
+) -> thread::JoinHandle<io::Result<()>> {
+    let local_inner = inner;
+    thread::spawn(move || -> io::Result<()> {
+        let mut running = true;
+        while running {
+            let inner = local_inner.lock().map_err(|e| {
+                io::Error::new(io::ErrorKind::Other, e.to_string())
+            })?;
+            let _span =
+                info_span!("running process", name = ?inner.name).entered();
+            match &inner.state {
+                ExecutableState::Init { .. } => {}
+                ExecutableState::Started { process, .. } => {
+                    let mut proc = process.lock().map_err(|e| {
```

Comment:
Won't this block `kill()` and `pid()` until there is nothing left to read from stdout/err? Is that ok?

Reply:
no. we don't read until "end", we just read whatever's currently in the pipe before unlocking.

it's possible that if the running executable is spamming stuff we might see a lock being held for longer.  maybe it's worth changing the process read to read a fixed amount (1k?) into a buffer, though then we might split log lines...